File: Tree/Medium/144-Binary-Tree-Preorder-Traversal/solution_iterative.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    public IList<int> PreorderTraversal(TreeNode root) {
        // iteration + stack
        // tc:O(n); sc:O(h)
        if (root == null) {
            return new List<int>();
        }
        
        var res = new List<int>();
        var stack = new Stack<TreeNode>();
        stack.Push(root);
        
        while (stack.Count > 0) {
            var node = stack.Pop();
            res.Add(node.val);
            
            // since stack is FILO, so put in right first, and it would be popped last
            if (node.right != null) {
                stack.Push(node.right);
            }
            
            if (node.left != null) {
                stack.Push(node.left);
            }
        }
        
        return res;
    }
    
}

File: Tree/Medium/144-Binary-Tree-Preorder-Traversal/solution_iterative.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# version 2: no adding a helper method
class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        # iteration + stack
        # tc:O(n); sc:O(h)
        
        if not root:
            return []
        
        res = []
        stack = [root]
        
        while stack:
            node = stack.pop()
            res.append(node.val)
            
            # since stack is FILO, so put in right first, and it would be popped last
            if node.right:
                stack.append(node.right)
                
            if node.left:
                stack.append(node.left)
            
        return res
        